backend/dal/Helpers/Extensions/EnumerableExtensions.cs

```diff
@@ -35,5 +37,25 @@ public static void ForEach<T>(this IEnumerable<T> items, Action<T> action)
 
             return item;
         }
+
+        /// <summary>
+        /// Provides a way to remove all items from the collection that match the find statement.
+        /// </summary>
+        /// <param name="items"></param>
+        /// <param name="find"></param>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        public static IEnumerable<T> RemoveAll<T>(this IEnumerable<T> items, Func<T, bool> find)
+        {
+            var list = items.ToList();
+            var remove = list.Where(find);
+
+            foreach (var item in remove)
+            {
+                list.Remove(item);
+            }
+
+            return list;
+        }
```

Comment:
I'd like to know why we are doing a custom "RemoveAll" here...
I understand that IEnumerable doesn't have it but List does - see [here](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.removeall?view=netcore-3.1)
We are already converting to a list here so I'd be tempted to go with:
```csharp
var list = items.ToList().RemoveAll(find)
```
Thoughts?

Reply:
Can go either way.
`items.ToList().RemoveAll(find);`
vs
`items.RemoveAll(find)`

Reply:
On second thoughts, if the list is large, converting to a list could impact performance. There seems to be an IEnumerable extension method already that can help us here:

```csharp
var itemsFiltered = items.SkipWhile(find);
```

See https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.skipwhile?view=netcore-3.1#definition

Reply:
That is an immutable solution.  The `RemoveAll(...)` mutates, which is the desired result in this case.

Reply:
Well, in my opinion this custom RemoveAll doesn't look very efficient. We are first iterating the whole list by doing `ToList()`. Then we are iterating it again with the `.Where()` so that's o(2N) and then we iterate the "remove" collection to remove from the original. 

The worst case scenario here - where ALL parcels need to be removed. We traverse the original list **3 times** - o(3N). That can degrade performance for big lists (ie a lot of parcels)

The "official" SkipWhile would iterate the list only once by asking each time: does this match the supplied predicate function? if yes, skip the element, if not leave it in the returning list

Reply:
About mutability, looking at the code below that uses your RemoveAll I don't see how it mutates the result....See my comments there

Reply:
The function is not performant.  But there isn't a good way to mutate an IEnumerable.  The only thing I could do is check if the original type passed is already a `List`.

That being said there are a number of possible other ways to do this.  I could change the entity and allow for the `Parcel.Buildings` collection to be `set`.  But this isn't ideal in-terms of a EF design.

Either way the extension method itself isn't supposed to be performant.  Nor could it be when mutating an item from the collection.

Reply:
I'm going to rewrite this extension to only support ICollection<T>, then it will do what I need it to do.  Which is mutate the collection.